MIGRATION_GUIDE.md

```diff
@@ -1,14 +1,264 @@
+
+`Handle::downcast` also requires a second type argument for the context type. This can usually be inferred, so you can typically use `_`.
+
+**Before:**
+```rust
+value.downcast::<JsNumber>()
+```
+
+**After:**
+
+```rust
+value.downcast::<JsNumber, _>(&mut cx)
+```
+
+##### Primitive types
+
+* `JsBoolean`
+    - `value`
+* `JsNull`
+    - `new`
 * `JsString`
     - `size`
     - `value`
+* `JsNumber`
+    - `value`
+* `JsUndefined`
+    - `new`
+
+##### Object APIs
+
 * `PropertyKey`
     - `get_from`
     - `set_from`
+
+### Handle equality
+
+Handles no longer implement `Eq` or `PartialEq`, which had underspecified behavior. Use `Value::strict_equals` instead to invoke the behavior of JavaScript's `===` operator.
+
+## Major API changes
+
+The N-API backend introduces two categories of significant change:
+
+1. Embedding Rust data, which is no longer done through the awkward and complex `declare_types!` (i.e. classes) macro, but through a simpler primitive: the `JsBox` API.
+2. Concurrency, which is offered through the Event Queue API instead of the Task API or Event Handlers, both of which are deprecated and removed in the N-API backend.
+
+### Embedding Rust data
+
+The `declare_types!` macro is deprecated and replaced by the `JsBox` type.
+
+_Rationale:_ The `declare_types!` macro provides a syntax for defining classes, but requires substantial boilerplate and is unergonomic for simple cases. It's also not flexible enough to express the full range of JavaScript classes syntax and semantics. With the `JsBox` type, it's easy to embed Rust data in JavaScript objects, which can then be nested inside of more feature-rich classes defined in pure JavaScript (or TypeScript).
+
+**Before:**
+
+```rust
+struct User {
+    first_name: String,
+    last_name: String,
+}
+
+impl User {
+    fn full_name(&self) -> String {
+        format!("{} {}", self.first_name, self.last_name)
+    }
+}
+
+declare_types! {
+    class JsUser for User {
+        init(mut cx) {
+            let first_name = cx.argument::<JsString>(0)?;
+            let last_name = cx.argument::<JsString>(1)?;
+            Ok(User { first_name, last_name })
+        }
+
+        method full_name(mut cx) {
+            let this = cx.this();
+            let guard = cx.lock();
+            let user = this.borrow(&guard);
+            let full_name = user.full_name();
+            Ok(cx.string(full_name).upcast())
+        }
+    }
+}
+```
+
+**After:**
+
+```rust
+struct User {
+    first_name: String,
+    last_name: String,
+}
+
+fn create_user(mut cx: FunctionContext) -> JsResult<JsBox<User>> {
+    let first_name = cx.argument::<JsString>(0)?.value(&mut cx);
+    let last_name = cx.argument::<JsString>(1)?.value(&mut cx);
+    Ok(cx.boxed(User { first_name, last_name }))
+}
+
+fn user_full_name(mut cx: FunctionContext) -> JsResult<JsString> {
+    let user = cx.argument::<JsBox<User>>(0)?;
+    let full_name = user.full_name();
+    Ok(cx.string(full_name))
+}
+```
```

Comment:
I think it would be helpful to include the JavaScript glue code as well to make this a fair comparison.

```js
class User {
    constructor(firstName, lastName) {
        self.native = addon.createUser(firstName, lastName);
    }
    
    fullName() {
        return addon.userFullName(this.native);
    }
}
```